### DISCUSSION TITLE
how to generate PKEY-SM2 and how to use SM2 encryption？

### DISCUSSION THREAD

**Discussion by iyouw** [openssl/openssl#25243] [Q&A] [answered]

**Comment by paulidale** [comment]
The existing demos should cover most of what you require: `demos/pkey` and `demos/signature`.  It shouldn't be much more than changing the algorithm name.

**Reply by iyouw** [reply]
i want to use sm2  encryption and decryption like `rsa` , not the signature. 
in the `demos/key/EVP_PKEY_EC_keygen` , should i replace the `EC` to `SM2` , or the `curve name` to `SM2`

**Reply by iyouw** [reply]
thank you very much for replying me, could you give me some demos?

**Reply by mattcaswell** [reply, marked as answer]
Like this:
````
    genctx = EVP_PKEY_CTX_new_from_name(libctx, "SM2", propq);
````

I don't think you need to set a curve name for SM2 IIRC. It just defaults to "SM2" I think.

**Reply by iyouw** [reply]
thank you very much for replying. i had successfully  generated the `PKEY-SM2` by `EVP_PKEY_CTX_new_from_name(libctx, "SM2", propq)`, and had successfully encoded the generated key by `OSSL_ENCODER_CTX_new_for_pkey` with key encoding parameter `PEM`. but when i decoded the `private key` by `OSSL_DECODER_CTX_new_for_pkey` with key encoding parameer `PEM`, key type parameter `SM2`, an error occurred. finally i found the same error at the issuse list.  so just waiting and hope new update will come soon!
